File: src/ast/graph.ts
import type { ParamValue, SourceLoc } from "./nodes.js";

/** Concrete shape dimensions, e.g. [3, 224, 224] */
export type Shape = number[];

/** A single computational node in the graph. */
export interface Block {
  /** Unique generated id, e.g. "block_0" */
  id: string;
  /** Layer type, e.g. "Conv2d" */
  type: string;
  params: Record<string, ParamValue>;
  /** Populated by shape inference */
  inputShapes: Shape[];
  /** Populated by shape inference; array supports multi-output blocks (e.g. LSTM: [hidden, cell]) */
  outputShapes: Shape[];
  /** Populated by shape inference; total parameter count for this block */
  paramCount?: number;
  /** Whether block height should scale with channel depth in SVG (from BlockDef.showDepth). Default: true. */
  showDepth?: boolean;
  loc: SourceLoc;
}

/** A directed data-flow edge between two blocks. */
export interface Edge {
  /** Block.id of the source block */
  from: string;
  /** Block.id of the destination block */
  to: string;
  /** Named tensor if this edge was created via `-> [name]` */
  tensorName?: string;
  /** Populated by shape inference */
  shape?: Shape;
}

/** A logical grouping / namespace for a set of blocks. */
export interface Group {
  /** Hierarchical path, e.g. ["ResNet50", "Bottleneck"] */
  path: string[];
  /** The ids of blocks belonging to this group */
  blockIds: string[];
}

/** The top-level directed graph IR produced after parsing. */
export interface Graph {
  blocks: Block[];
  edges: Edge[];
  groups: Group[];
}

File: mlmd-core/src/ast/graph.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::ast::nodes::{ParamValue, SourceLoc};

// ---------------------------------------------------------------------------
// Shape type alias
// ---------------------------------------------------------------------------

pub type Shape = Vec<usize>;

// ---------------------------------------------------------------------------
// Block — a single computational node in the graph
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Block {
    /// Unique generated id, e.g. "block_0"
    pub id: String,
    /// Layer type, e.g. "Conv2d"
    pub block_type: String,
    /// Resolved parameters
    pub params: HashMap<String, ParamValue>,
    /// Populated by shape inference
    pub input_shapes: Vec<Shape>,
    /// Populated by shape inference; array supports multi-output blocks
    pub output_shapes: Vec<Shape>,
    /// Populated by shape inference; total parameter count for this block
    pub param_count: Option<usize>,
    /// Whether block height should scale with channel depth in SVG
    pub show_depth: Option<bool>,
    /// Source location from the original source text
    pub loc: SourceLoc,
}

// ---------------------------------------------------------------------------
// Edge — a directed data-flow edge between two blocks
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Edge {
    /// Block.id of the source block
    pub from: String,
    /// Block.id of the destination block
    pub to: String,
    /// Named tensor if this edge was created via `-> [name]`
    pub tensor_name: Option<String>,
    /// Populated by shape inference
    pub shape: Option<Shape>,
}

// ---------------------------------------------------------------------------
// Group — a logical grouping / namespace for a set of blocks
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Group {
    /// Hierarchical path, e.g. ["ResNet50", "Bottleneck"]
    pub path: Vec<String>,
    /// The ids of blocks belonging to this group
    pub block_ids: Vec<String>,
}

// ---------------------------------------------------------------------------
// Graph — top-level directed graph IR produced after parsing
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Graph {
    pub blocks: Vec<Block>,
    pub edges: Vec<Edge>,
    pub groups: Vec<Group>,
}
